### ISSUE TITLE
Second Pod Using Same PVC Fails WaitForAttach Flakily

### ISSUE THREAD

**Issue by davidz627** [kubernetes/kubernetes#65246]
The test `[sig-storage] CSI Volumes CSI plugin test using CSI driver: hostPath should provision storage` flakes very occasionally. Below are the findings of the circumstances of the flake.

During `testDynamicProvisioning` two pods are created in quick succession. The test flakes when the second Pod fails to come up.

The second pod fails to come up with error:
`MountVolume.WaitForAttach failed for volume "pvc-bc2294b873ef11e8" : resource name may not be empty`

Upon exploring the logs we discovered that there was only 1 Attach call made at the beginning, and one Detach call upon timeout of the test [See timestamps]:
```
I0613 19:00:41.888473       1 reconciler.go:291] attacherDetacher.AttachVolume started for volume "pvc-0c896e426f3c11e8" (UniqueName: "kubernetes.io/csi/csi-hostpath^0da20430-6f3c-11e8-a80a-0a580a40016c") from node "e2e-40645-674b9-minion-group-gz10" 
I0613 19:00:42.533343       1 operation_generator.go:335] AttachVolume.Attach succeeded for volume "pvc-0c896e426f3c11e8" (UniqueName: "kubernetes.io/csi/csi-hostpath^0da20430-6f3c-11e8-a80a-0a580a40016c") from node "e2e-40645-674b9-minion-group-gz10" 
I0613 19:15:52.528902       1 reconciler.go:233] attacherDetacher.DetachVolume started for volume "pvc-0c896e426f3c11e8" (UniqueName: "kubernetes.io/csi/csi-hostpath^0da20430-6f3c-11e8-a80a-0a580a40016c") on node "e2e-40645-674b9-minion-group-gz10" 
I0613 19:15:52.615262       1 operation_generator.go:1225] Verified volume is safe to detach for volume "pvc-0c896e426f3c11e8" (UniqueName: "kubernetes.io/csi/csi-hostpath^0da20430-6f3c-11e8-a80a-0a580a40016c") on node "e2e-40645-674b9-minion-group-gz10" 
I0613 19:15:53.367487       1 operation_generator.go:419] DetachVolume.Detach succeeded for volume "pvc-0c896e426f3c11e8" (UniqueName: "kubernetes.io/csi/csi-hostpath^0da20430-6f3c-11e8-a80a-0a580a40016c") on node "e2e-40645-674b9-minion-group-gz10" 
```

We tracked down the `resource name may not be empty` error to the likely culprit in func `waitForVolumeAttachmentInternal`:
```
attach, err := c.k8s.StorageV1beta1().VolumeAttachments().Get(attachID, meta.GetOptions{})
```
based on the error message returned. 

This is likely caused by `attachID`(string) being empty [uncomfirmed], which is what the Kubelet would get from the DevicePath field on the Node object. This DevicePath field is normally added on Attach and removed on Detach of the volume.

The DevicePath is clearly present on the Node when the first pod is being brought up. But it is possible that it is somehow missing or not able to be retrieved by the time the second Pod is starting up (and continues to be unavailable exhausting the 15m timeout on the test) even though no Detach operation was performed on the volume in the meantime (according to logs). We are currently unable to reproduce this flake on command and did all analysis based on logs from the two flakes we saw naturally occur in the 100s of runs performed of the test.

This flake was originally seen in #65064 and the test logs can be found here: https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/65064/pull-kubernetes-e2e-gce/40645/

**Environment**: GCE
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"8", GitVersion:"v1.8.6", GitCommit:"6260bb08c46c31eea6cb538b34a9ceb3e406689c", GitTreeState:"clean", BuildDate:"2017-12-21T06:34:11Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"12+", GitVersion:"v1.12.0-alpha.0.975+98194b8630958e", GitCommit:"98194b8630958ee48db662f42404dc0bc6a2ed80", GitTreeState:"clean", BuildDate:"2018-06-19T18:16:46Z", GoVersion:"go1.10.2", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration: GCP
- OS (e.g. from /etc/os-release):
```
BUILD_ID=10323.64.0
NAME="Container-Optimized OS"
KERNEL_COMMIT_ID=94bdaa2af49562e7aba0168cae050911d105f5cb
GOOGLE_CRASH_ID=Lakitu
VERSION_ID=65
BUG_REPORT_URL=https://crbug.com/new
PRETTY_NAME="Container-Optimized OS from Google"
VERSION=65
GOOGLE_METRICS_PRODUCT_ID=26
HOME_URL="https://cloud.google.com/compute/docs/containers/vm-image/"
ID=cos
```
- Kernel (e.g. `uname -a`): `Linux e2e-test-dyzz-master 4.4.111+ #1 SMP Wed Mar 28 13:44:38 PDT 2018 x86_64 Intel(R) Xeon(R) CPU @ 2.30GHz GenuineIntel GNU/Linux`

/kind flake
/kind bug
/sig storage
/cc @msau42 @jingxu97 @saad-ali @verult

**Comment by jberkus** [comment]
Hey, is this test still flaking?

**Comment by jberkus** [comment]
/remove-lifecycle stale

A check in triage shows that this failing test is still a problem, and in fact has become more of a problem since October 4th.  Has there been any progress on a fix?

https://storage.googleapis.com/k8s-gubernator/triage/index.html?pr=1&test=hostPath%20should%20provision%20storage

**Comment by saad-ali** [comment]
We should find the root cause and fix for 1.13

**Comment by saad-ali** [comment]
/assign @pohly 
Patrick can you PTAL at what is causing flakiness in this test.

**Comment by pohly** [comment]
I will have a look. Right now the log output of the sidecar containers and the CSI driver are not captured, which makes debugging a bit hard. In PR #69868 I am adding code which includes those message in the test log. Let me finish that PR and once it is merged, debugging the next test flake might provide some better insight into what went wrong.

**Comment by jberkus** [comment]
@pohly @saad-ali

Hey, this issue is considered a blocker for 1.13beta.  Is there anything we can do to get it resolved sooner, so that it doesn't delay beta next week?

**Comment by msau42** [comment]
@pohly is trying to instrument the e2e tests to get logs from the csi driver containers so that we can debug.  He has tried running the test locally ~100 times and didn't see it flake.

**Comment by msau42** [comment]
@pohly here's one recent one you can check out:
https://gubernator.k8s.io/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-multizone/25313#sig-storage-csi-volumes-csi-plugin-test-using-csi-driver-hostpath-should-provision-storage

**Comment by pohly** [comment]
Michelle Au <notifications@github.com> writes:

> @pohly here's one recent one you can check out:
> https://gubernator.k8s.io/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-multizone/25313#sig-storage-csi-volumes-csi-plugin-test-using-csi-driver-hostpath-should-provision-storage

I had a look, but something is odd about that one. The error for the
second pod is there:
```
Nov  5 00:01:27.206: INFO: At 2018-11-04 23:46:24 +0000 UTC - event for
pvc-volume-tester-8lxcr: {kubelet
bootstrap-e2e-gce-mz-minion-group-3kn6} FailedMount:
MountVolume.WaitForAttach failed for volume
"pvc-cb81bf9f-e08b-11e8-857a-42010a800002" : volume
cf5aa27a-e08b-11e8-af35-42010a800007 has GET error for volume attachment
: resource name may not be empty
```

But under
http://gcsweb.k8s.io/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-multizone/25313/artifacts/
there is no further information about node
bootstrap-e2e-gce-mz-minion-group-3kn6. Shouldn't that be available?

**Comment by pohly** [comment]
"Patrick Ohly" <patrick.ohly@intel.com> writes:

> Michelle Au <notifications@github.com> writes:
> But under
> http://gcsweb.k8s.io/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-multizone/25313/artifacts/
> there is no further information about node
> bootstrap-e2e-gce-mz-minion-group-3kn6. Shouldn't that be available?

Here's a better example where the kublet log with the error is
available:
https://gubernator.k8s.io/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce/31590#sig-storage-csi-volumes-csi-plugin-test-using-csi-driver-hostpath-should-provision-storage

That one also has the log of the external-attacher, but that doesn't
tell me much.

I'm now comparing a "bad" run with a "good" local run. One observation
is that in the "good" run, operationExecutor.UnmountDevice is completed
before the second usage, including the "Volume detached for volume" step
in reconciler.go. Then operationExecutor.VerifyControllerAttachedVolume
is done again for the second pod, exactly as it was done for the first:

```
I1105 15:05:05.028142  345635 reconciler.go:301] Volume detached for volume "pvc-bfa311c0-e103-11e8-be03-fcaa1497a416" (UniqueName: "kubernetes.io/csi/csi-hostpath-e2e-tests-csi-volumes-59q8f^c4855a3c-e103-11e8-8f99-fcaa1497a416") on node "127.0.0.1" DevicePath ""
I1105 15:05:05.228621  345635 reconciler.go:207]
operationExecutor.VerifyControllerAttachedVolume started for volume
"pvc-bfa311c0-e103-11e8-be03-fcaa1497a416" (UniqueName:
"kubernetes.io/csi/csi-hostpath-e2e-tests-csi-volumes-59q8f^c4855a3c-e103-11e8-8f99-fcaa1497a416")
pod "pvc-volume-tester-p6fsj" (UID:
"cb9594d9-e103-11e8-be03-fcaa1497a416"
```

In the "bad" run, operationExecutor.MountVolume is invoked before that
"Volume detached for volume" and then runs with an empty device path,
without going through VerifyControllerAttachedVolume again:

```
I1105 04:41:19.118304    1582 reconciler.go:294] operationExecutor.UnmountDevice started for volume "pvc-00f43625-e0b5-11e8-a0e7-42010a800002" (UniqueName: "kubernetes.io/csi/csi-hostpath-e2e-tests-csi-volumes-rl75x^05fcbadb-e0b5-11e8-9ff2-42010a800006") on node "bootstrap-e2e-minion-group-l5mk"
...
I1105 04:41:19.123352    1582 desired_state_of_world_populator.go:318] Added volume "my-volume" (volSpec="pvc-00f43625-e0b5-11e8-a0e7-42010a800002") for pod "09937776-e0b5-11e8-a0e7-42010a800002" to desired state.
...
I1105 04:41:19.143198    1582 operation_generator.go:768] UnmountDevice
succeeded for volume "pvc-00f43625-e0b5-11e8-a0e7-42010a800002" %!(EXTRA
string=UnmountDevice succeeded for volume
"pvc-00f43625-e0b5-11e8-a0e7-42010a800002" (UniqueName:
"kubernetes.io/csi/csi-hostpath-e2e-tests-csi-volumes-rl75x^05fcbadb-e0b5-11e8-9ff2-42010a800006")
on node "bootstrap-e2e-minion-group-l5mk" )
...
I1105 04:41:19.228716    1582 reconciler.go:252] operationExecutor.MountVolume started for volume "pvc-00f43625-e0b5-11e8-a0e7-42010a800002" (UniqueName: "kubernetes.io/csi/csi-hostpath-e2e-tests-csi-volumes-rl75x^05fcbadb-e0b5-11e8-9ff2-42010a800006") pod "pvc-volume-tester-jq4lz" (UID: "09937776-e0b5-11e8-a0e7-42010a800002")
I1105 04:41:19.228869    1582 operation_generator.go:497] MountVolume.WaitForAttach entering for volume "pvc-00f43625-e0b5-11e8-a0e7-42010a800002" (UniqueName: "kubernetes.io/csi/csi-hostpath-e2e-tests-csi-volumes-rl75x^05fcbadb-e0b5-11e8-9ff2-42010a800006") pod "pvc-volume-tester-jq4lz" (UID: "09937776-e0b5-11e8-a0e7-42010a800002") DevicePath ""
```

So to me it looks like a race between two conflicting operations. Does
that make sense?

How to fix it is less obvious. The relevant code seems to be this in reconciler.go:
```
	// Ensure volumes that should be attached/mounted are attached/mounted.
	for _, volumeToMount := range rc.desiredStateOfWorld.GetVolumesToMount() {
		volMounted, devicePath, err := rc.actualStateOfWorld.PodExistsInVolume(volumeToMount.PodName, volumeToMount.VolumeName)
		volumeToMount.DevicePath = devicePath
		if cache.IsVolumeNotAttachedError(err) {
			if rc.controllerAttachDetachEnabled || !volumeToMount.PluginIsAttachable {
				// Volume is not attached (or doesn't implement attacher), kubelet attach is disabled, wait
				// for controller to finish attaching volume.
				glog.V(5).Infof(volumeToMount.GenerateMsgDetailed("Starting
                                operationExecutor.VerifyControllerAttachedVolume", ""))

        ...

	// Ensure devices that should be detached/unmounted are detached/unmounted.
	for _, attachedVolume := range rc.actualStateOfWorld.GetUnmountedVolumes() {
		// Check IsOperationPending to avoid marking a volume as detached if it's in the process of mounting.
		if !rc.desiredStateOfWorld.VolumeExists(attachedVolume.VolumeName) &&
			!rc.operationExecutor.IsOperationPending(attachedVolume.VolumeName, nestedpendingoperations.EmptyUniquePodName) {
			if attachedVolume.GloballyMounted {
				// Volume is globally mounted to device, unmount it
				glog.V(5).Infof(attachedVolume.GenerateMsgDetailed("Starting operationExecutor.UnmountDevice", ""))
				err := rc.operationExecutor.UnmountDevice(
					attachedVolume.AttachedVolume, rc.actualStateOfWorld, rc.mounter)
				if err != nil &&
					!nestedpendingoperations.IsAlreadyExists(err) &&
					!exponentialbackoff.IsExponentialBackoff(err) {
					// Ignore nestedpendingoperations.IsAlreadyExists and exponentialbackoff.IsExponentialBackoff errors, they are expected.
					// Log all other errors.
					glog.Errorf(attachedVolume.GenerateErrorDetailed(fmt.Sprintf("operationExecutor.UnmountDevice failed (controllerAttachDetachEnabled %v)", rc.controllerAttachDetachEnabled), err).Error())
				}
				if err == nil {
					glog.Infof(attachedVolume.GenerateMsgDetailed("operationExecutor.UnmountDevice started", ""))
				}
			} else {
				// Volume is attached to node, detach it
				// Kubelet not responsible for detaching or this volume has a non-attachable volume plugin.
				if rc.controllerAttachDetachEnabled || !attachedVolume.PluginIsAttachable {
					rc.actualStateOfWorld.MarkVolumeAsDetached(attachedVolume.VolumeName, attachedVolume.NodeName)
					glog.Infof(attachedVolume.GenerateMsgDetailed("Volume detached", fmt.Sprintf("DevicePath %q", attachedVolume.DevicePath)))

```

@msau42 @saad-ali Any suggestions?

**Comment by pohly** [comment]
Michelle Au <notifications@github.com> writes:
> This seems inline with the observations that @davidz627 made
> initially. We should trace where DevicePath is getting set.  Could
> there be a race with AD controller or external attacher if they try to
> detach the volume?

No idea.

> https://github.com/kubernetes/kubernetes/issues/70319 is another race
> around this area. Unclear if it's the same yet.

Even if it isn't, it's at least related to the same piece of
code. Probably worth fixing both together.

Can we get someone who has worked in this area to fix these issues?
Today is the first day that I've ever seen this code, so I'm a bit
afraid that it'll take me a long time to get it right, and I'm worried
about regressions.

**Comment by msau42** [comment]
I've asked @gnufied to help investigate

**Comment by saad-ali** [comment]
It is not a blocker for beta or freeze. Can be addressed in stabilization. 
Let's keep the 1.13 tag on it. And based on progress by stabilization we can reassess if it is even a 1.13 blocker (at this point I don't believe it is).

**Comment by pohly** [comment]
The `devicePath` in the actual state of the world gets reset during unmounting:
```
(dlv) p volumeObj
k8s.io/kubernetes/pkg/kubelet/volumemanager/cache.attachedVolume {
	volumeName: "kubernetes.io/csi/csi-hostpath-e2e-tests-csi-volumes-vprlp^0d236...+31 more",
	mountedPods: map[k8s.io/kubernetes/pkg/volume/util/types.UniquePodName]k8s.io/kubernetes/pkg/kubelet/volumemanager/cache.mountedPod [],
	spec: *k8s.io/kubernetes/pkg/volume.Spec {
		Volume: *k8s.io/kubernetes/vendor/k8s.io/api/core/v1.Volume nil,
		PersistentVolume: *(*k8s.io/kubernetes/vendor/k8s.io/api/core/v1.PersistentVolume)(0xc420f95400),
		ReadOnly: false,},
	pluginName: "kubernetes.io/csi",
	pluginIsAttachable: true,
	globallyMounted: false,
	devicePath: "",
	deviceMountPath: "",}
(dlv) bt
0  0x0000000005763350 in k8s.io/kubernetes/pkg/kubelet/volumemanager/cache.(*actualStateOfWorld).SetVolumeGloballyMounted
   at ./pkg/kubelet/volumemanager/cache/actual_state_of_world.go:573
1  0x00000000057602d6 in k8s.io/kubernetes/pkg/kubelet/volumemanager/cache.(*actualStateOfWorld).MarkDeviceAsUnmounted
   at ./pkg/kubelet/volumemanager/cache/actual_state_of_world.go:358
2  0x0000000004717ec4 in k8s.io/kubernetes/pkg/volume/util/operationexecutor.(*operationGenerator).GenerateUnmountDeviceFunc.func1
   at ./pkg/volume/util/operationexecutor/operation_generator.go:771
3  0x0000000004701ab6 in k8s.io/kubernetes/pkg/volume/util/nestedpendingoperations.(*nestedPendingOperations).Run.func1
   at ./pkg/volume/util/nestedpendingoperations/nestedpendingoperations.go:143
4  0x0000000000464db1 in runtime.goexit
   at /usr/lib/go-1.10/src/runtime/asm_amd64.s:2361
```

**Comment by jberkus** [comment]
@saad-ali 

The RT considers this a blocker now, because flakes are happening often enough to interfere with getting clear signal on several jobs.  Even if the underlying problem isn't that serious.

**Comment by msau42** [comment]
/assign
I will help triage as well.

**Comment by saad-ali** [comment]
Ack @jberkus. Thanks @msau42

**Comment by saad-ali** [comment]
@pohly Instead of https://github.com/kubernetes/kubernetes/pull/70746 which may have negative side effects for non-CSI volumes, let's do this instead:

* Modify the CSI `WaitForAttach(...)` method in https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/csi/csi_attacher.go#L126 to ignore the passed in `attachID string` (which is populated from devicePath. Instead let's have that method generate the `attachId` from scratch, e.g. `	attachID := getAttachmentName(volID, driverName, string(nodeName))
`

This will remove any CSI dependency on devicePath and not effect other volume plugins.

**Comment by jsafrane** [comment]
I am trying to twist my head around the code. IMO it affects also AWS that uses devicePath stored in node.status and it should be reproducible there too. So far I was unsuccessful.

In the logs it's clearly visible that UnmountDevice succeeded. So why kubelet continues with MountVolume instead of MountDevice? That's wrong. A workaround in CSI volume plugin won't really help here, it will work only for drivers without `NodeStageVolume`.

**Comment by jsafrane** [comment]
I finally reproduced it on AWS with a long sleep in `awsElasticBlockStoreDetacher.UnmountDevice`:

```go
	klog.Infof("JSAF: slowing down UnmountDevice")
	time.Sleep(time.Minute)
	klog.Infof("JSAF: UnmountDevice finished")
	return volumeutil.UnmountPath(deviceMountPath, detacher.mounter)
```

Steps to reproduce (with the hack above):

1. Start a pod with a volume
2. Delete the pod. Wait until `UnmountDevice` is called.
3. Start a new pod with the same volume.

-> [reconciler](https://github.com/kubernetes/kubernetes/blob/db680631df67c1d2592fba2edd2823522f105b12/pkg/kubelet/volumemanager/reconciler/reconciler.go#L187) checks that the volume is still attached and the device is still mounted and enqueues an operation to mount the volume.

When `UnmountDevice` finishes, `MountVolume` starts and 1) uses empty `devicePath` and 2) calls `SetUpAt` on volume that's not DeviceMounted.

**Comment by jsafrane** [comment]
Fix: https://github.com/kubernetes/kubernetes/pull/71074

Any idea how to test it for real? Fake CSI plugin with slow NodeUnstage would be the best, however we don't have one.

**Comment by msau42** [comment]
I'm going to work on a targeted fix for CSI.  We can continue to pursue fixing aws in-tree separately.

**Comment by msau42** [comment]
PR to fix csi is here: https://github.com/kubernetes/kubernetes/pull/71095

**Comment by msau42** [comment]
Only #71095 is needed to fix this particular test failure.

**Comment by AishSundar** [comment]
@jsafrane @msau42 can you please confirm if we are going to fix this using #71074 or #71095?

**Comment by msau42** [comment]
I discussed with @jsafrane offline, he has some suggestions for #71095 that I am investigating.  I think we will want both PRs eventually, although only one of them will be needed to fix the CSI test for now.  Since it's his off hours, I will continue working on #71095 today.

**Comment by AishSundar** [comment]
``` I think we will want both PRs eventually```
Are we talking 1.13 for #71074 as well?

**Comment by AishSundar** [comment]
Reopening this since we are seeing consistent failures in latest runs. This is now 1.13 release blocking

https://k8s-testgrid.appspot.com/sig-release-master-upgrade#gce-new-master-upgrade-cluster

https://gubernator.k8s.io/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-new-master-upgrade-cluster/2629

@msau42 @jsafrane

**Comment by jberkus** [comment]
@msau42 "intentionally not backwards compatible" -- is there a writeup for the release notes and release announcement for this change?

**Comment by msau42** [comment]
Yes https://github.com/kubernetes/kubernetes/pull/71020 has the release note

**Comment by jberkus** [comment]
@msau42 thanks, I'll also call it to the Release Notes lead's attention

**Comment by AishSundar** [comment]
@saad-ali @msau42 what does it mean from the user impact perspective for CSI to not be backward compatible with 1.12? can you please point me to any discussions on this behavior for 1.13?

**Comment by davidz627** [comment]
Im tracking these upgrade test failures elsewhere: https://github.com/kubernetes/kubernetes/issues/71228
They are not related to this original issue, the flaky test referenced by this issue should have been solved by #71074
/close

**Comment by AishSundar** [comment]
@dstrebel @marpaia as FYI to call this out clearly in the Release notes.
